File: engines/py/anza/engine/slot/parse.py
from dataclasses import dataclass
from typing import Literal
from ...errors import AnzaError


@dataclass(frozen=True)
class Chunk:
    kind: Literal["static", "slot"]
    value: str


def extract(template: str) -> list[Chunk]:
    chunks: list[Chunk] = []
    cursor = 0
    length = len(template)

    while cursor < length:
        open_idx = template.find("{{", cursor)
        if open_idx == -1:
            chunks.append(Chunk(kind="static", value=template[cursor:]))
            break

        if open_idx > cursor:
            chunks.append(Chunk(kind="static", value=template[cursor:open_idx]))

        close_idx = template.find("}}", open_idx + 2)
        if close_idx == -1:
            raise AnzaError.template("Unclosed slot placeholder '{{' in template")

        slot_name = template[open_idx + 2 : close_idx].strip()
        if not slot_name:
            raise AnzaError.template("Empty slot placeholder {{}} in template")

        chunks.append(Chunk(kind="slot", value=slot_name))
        cursor = close_idx + 2

    return chunks

File: engines/rust/src/engine/slot/parse.rs
use {
  crate::errors::{Error, Result},
};

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Chunk {
  Static(Vec<u8>),
  Slot(String),
}

pub fn extract(template: &str) -> Result<Vec<Chunk>> {
  let mut chunks = Vec::new();
  let mut cursor = 0;
  let bytes = template.as_bytes();

  while let Some(open_idx) = template[cursor..].find("{{") {
    let start = cursor + open_idx;
    if start > cursor {
      chunks.push(Chunk::Static(bytes[cursor..start].to_vec()));
    }

    if let Some(close_idx) = template[start..].find("}}") {
      let end = start + close_idx;
      let slot_name = template[start + 2..end].trim().to_string();
      if slot_name.is_empty() {
        return Err(Error::template("Empty slot placeholder {{}} in template"));
      }
      chunks.push(Chunk::Slot(slot_name));
      cursor = end + 2;
    } else {
      return Err(Error::template("Unclosed slot placeholder '{{' in template"));
    }
  }

  if cursor < bytes.len() {
    chunks.push(Chunk::Static(bytes[cursor..].to_vec()));
  }

  Ok(chunks)
}
